WIP: balance-printing

vindard opened the pull request:
Issues:
1. Balance was getting reported in '/tip' unwantedly, balance formatting wasn't ideal.
2. No message was being returned after '/withdraw' call

Result:
Both issues were addressed

Add'l:
1. A DGE/USD rate fetcher was added
2. Context was added to the 'address' messages by adding a link to an external Doge Explorer.

vindard commented:
I just discovered some new things about the withdraw method from the blockio API. I'm closing this to add some updates to '/withdraw' and I'll reopen once I'm finished.

johnsBeharry commented on run.py:
```diff
@@ -51,8 +69,8 @@ def process(message,username,chatid):
-			(balance, pending_balance) = returnBal(username)
-			sendMsg("@"+username+" Balance : "+balance+ "Doge ("+pending_balance+" Doge)",chatid)
+			(balance, pending_balance, balance_msg, pending_msg) = returnBal()
+			sendMsg("@"+'vindard'+balance_msg+pending_msg,chatid)
```
@vindard i think this was left behind from testing - it will return your username each time `/balance` is sent.

vindard replied:
Ah good catch. Just fixed.

vindard commented:
Ok '/withdraw' messages fixed up. Should be pretty intuitive now hopefully.

johnsBeharry commented:
@vindard first off thanks for the work in this PR. There's some difficulty reviewing it though as I'm not sure what acceptance criteria/scope I should be testing against. Can you provide some clarification around the specific areas you need tested?

I ran into a few problem with the bot not being responsive to the `/balance` command, and registration not working with my username, I suspect the latter problem may be outside the scope of the changes you made but having that clarified would help investigating the problem.


---

This seems like its' a good time to start setting up some testing framework. I lean towards Behavior Driven Development (BDD) style although in most open source projects its typical to use traditional Test Driven Development with a unit testing framework.

vindard commented:
@johnsBeharry hey, hmm that's strange. Ok cool I'll check out that link to see if I could better explain what to look out for. Thanks man, I've actually been wondering how to better communicate these things I'm changing. 

Looks like I really need to work on getting the bot set up locally as well. There shouldn't be any issues with the /register and /balance in this PR, that's very strange. The things I changed were only within the sendMsg() part of the /balance call, so even if something was broken it should default to the exception case and return that "you are not registered" message.

I'll review as well on my end once I get the bot up and running and let you know ;)

vindard commented:
@johnsBeharry hey ok I think I cleaned up a bit in this PR. I'm going to take a stab at explaining what to look for on testing.

What I did in this PR was to clean up how messages are returned to sender after the following calls: '/balance', '/address' and '/withdraw'.

'/address' was pretty straightforward. I just added a one-liner to link to block explorer.

For '/balance' and '/withdraw', I put the actual messages for these in functions higher up and had the messages be returned as outputs from those functions. Then I simply assigned those outputs within the "def process" logic to get the messages that would be sent by the bot to the user.

The testing should mostly be just making sure the right messages are returned when those three commands are called.

Hope this helps 😉

micey969 commented:
hey @vindard hey i was testing out your 1c tipping branch pull request. i was trying the withdraw function. in what scenarios would i get the other withdraw msg besides the success one and the insufficient balance/not registered one?

micey969 commented:
Also, one last thing, can you remove your getCount function inside of run.py to match with the Develop branch. Also, can you change the instances of it where it is called in run.py to getActive. I had changed it to getActive when I did my first merge because it was a more appropriate name of what the function did. You can look at the code for the develop branch to see what I'm referring to.
Once that's done, I will merge it. Also, please do a pull before you work on it, there was a conflict in the branch that i resolved. Thanks

vindard commented:
@micey969 ok great I believe all those things you mentioned should be fixed now. Thanks a lot for reviewing, and let me know if there's anything else that's still not quite right with this branch.
